apps/opencs/view/render/worldspacewidget.hpp

```diff
@@ -35,6 +36,14 @@ namespace CSVRender
     class CellArrow;
     class EditMode;
 
+    struct WorldspaceHitResult
+    {
+        bool hit;
+        osg::ref_ptr<TagBase> tag;
+        unsigned int i0, i1, i2;
+        osg::Vec3d worldPos;
+    };
```

Comment:
The tag based system is good, but does not pass enough information. Since I implemented the rendering as a single object, I need triangle data (the poorly named i0,i1,i2 variables) to properly select nodes. I also added the world position since it will probably be convenient and has already been found. This will be needed to add a pathgrid node at a terrain position.
The alternatives I can think of are to pass that data in the tag itself (seems to not fit a tags purpose though) or slightly restructure the worldspace widget so that it passes the mouse coordinates to the edit mode which can then decide how to use it.